Basic complex action implementation

XRater opened the pull request.

XRater committed:
Basic complex action implementation

vasyoid commented on core/src/main/kotlin/ru/yandex/testopithecus/monkeys/complex/complex_action/splitter/SimpleComplexActionSplitter.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    override fun pushComplexAction(action: ComplexAction) {
+        complexAction = Optional.of(action)
```
Хорошо бы здесь добавить проверку на то, что мы полностью исчерпали предыдущий экшн

vasyoid commented on core/src/main/kotlin/ru/yandex/testopithecus/monkeys/complex/complex_action/splitter/SimpleComplexActionSplitter.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private fun updateOptionalIfEmpty() {
+        if (index == complexAction.get().actionDescriptions.size) {
```
И здесь тоже хорошо бы проверить, что complexAction не пустой. Иначе исключение вылетит

vasyoid commented on core/src/main/kotlin/ru/yandex/testopithecus/monkeys/complex/complex_action/splitter/SimpleComplexActionSplitter.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private fun getIdForAction(elementDescription: ElementDescription, actionDescription: ActionDescription, state: UiState): String? {
+        if (actionDescription.attributes["global"] == "true") {
```
Можно то что экшн глобальный, как-то в ElementDesctiption отражать. Например, делать attributes пустыми или вообще null или сам elementDescription = null

vasyoid commented on core/src/main/kotlin/ru/yandex/testopithecus/monkeys/complex/ComplexMonkey.kt:
```diff
@@ -0,0 +1,39 @@
+    override fun generateAction(uiState: UiState): UiAction {
+        if (complexActionSplitter.isEmpty()) {
+            val nextAction = baseMonkey.generateAction(uiState)
+            val simpleAction = parseAction(nextAction, uiState) // Temporary
```
Temporary относится вообще ко всему parseAction()? Потом он удалится?

XRater replied:
Да, потом из `Monkey` будет сразу возвращаться `ComplexAction`

vasyoid reviewed (commented):
В целом выглядит хорошо. Если заработает и будет хорошо искать похожие элементы, будет круто

XRater committed:
Minor fixes after review

1) Added empty checks for SimpleComplexActionSlitter
2) Global action now has null ElementDescription

XRater committed:
Merge branch 'master' into complex-action

# Conflicts:
#	app/src/androidTest/java/ru/yandex/testopithecus/system/AndroidMonkey.kt
